backends/p4tools/common/lib/symbolic_env.cpp

```diff
@@ -28,7 +28,18 @@ const IR::Expression *SymbolicEnv::get(const IR::StateVariable &var) const {
 bool SymbolicEnv::exists(const IR::StateVariable &var) const { return map.find(var) != map.end(); }
 
 void SymbolicEnv::set(const IR::StateVariable &var, const IR::Expression *value) {
-    map[var] = P4::optimizeExpression(value);
+    const auto *type = value->type;
+    BUG_CHECK(type && !type->is<IR::Type_Unknown>(), "Cannot set value with unspecified type: %1%",
+              value);
+    value = P4::optimizeExpression(value);
+    // The P4 passes used in optimizeExpression can strip types which then breaks P4Tools, make sure
+    // at least the top-level type is set.
+    if (value->type->is<IR::Type_Unknown>()) {
```

Comment:
I'd prefer to fix the optimization pass and have a BUG_CHECK here instead. We are using StrengthReduction and ConstantFolding. They definitely should preserve types.

Reply:
I will have a look into those two passes. But in general my impression from many of the P4 passes in the compiler is that they don't bother with types and just expect type checker to run after :-/. I think that is conceptually a wrong decision, but it could be hard to change it.

Reply:
True, but we can still fix up individual passes that only handle expressions etc. I think I stumbled across this problem before and did some work to fix it. The changes necessary to the passes were not major but I unfortunately discarded them..

Reply:
@fruffy, I came across one point where the optimizer does change the type. When there is just a varbit extracted (e.g. `testgen-p4c-bmv2-protobuf-ir/issue447-bmv2.p4`), the packet expression is something line `0 ++ (SizedVarbit<32><16>)|pktvar_0(bit<16>)|` which is of type `bit<16>` but the optimizer turns it into `(SizedVarbit<32><16>)|pktvar_0(bit<16>)|` which is `SizedVarbit<32><16>`. For now I've commited a change that considers `bit<N>` and `SizedVarbit<M><N>` the same for the validity check.

Looking at the other code that relates to varbits, it seems like these are mostly special cases that treat `SizedVarbit<M><N>` as `bit<N>`. So I think we might be just fine leaving the `varbit<M>` in the IR and handling the special case in `extract` only (with the extracted type being `bit<N>`. This would probably simplify the overall handling. The only cost I can see would be that the trace would not contain info that the type is a varbit. What do you think? This should be also OK for the targets here in p4c, I'm not sure about Tofino (@pkotikal) and of course other targets that I don't know about. But I think that should be a separate PR.

Reply:
Let's raise an issue for this so it is not forgotten but leave it for later (I'll raise the issue).